File: js/6.js
#!/usr/bin/env node
'use strict'

const fs = require('fs')

const parse = (str) => str.split('\n')
  .filter(x => x.length > 0)
  .map((str) => {
    const parts = str.match(/(.*) (\d+),(\d+) through (\d+),(\d+)/)
    return {
      action: parts[1],
      start: {
        x: Math.min(parts[2], parts[4]),
        y: Math.min(parts[3], parts[5])
      },
      end: {
        x: Math.max(parts[2], parts[4]),
        y: Math.max(parts[3], parts[5])
      }
    }
  })

const sideA = (arr) => {
  const followInstruction = (grid, instruction) => {
    for (let x = instruction.start.x; x <= instruction.end.x; ++x) {
      for (let y = instruction.start.y; y <= instruction.end.y; ++y) {
        if (grid[x] === undefined) grid[x] = []
        // if (grid[x][y] === undefined) grid[x][y] = false
        switch (instruction.action) {
          case 'turn on':
            grid[x][y] = true
            break
          case 'turn off':
            grid[x][y] = false
            break
          case 'toggle':
            grid[x][y] = grid[x][y] ? !grid[x][y] : true
            break
        }
      }
    }
    return grid
  }

  return arr.reduce(followInstruction, [])
    .reduce((count, row) => count + row.filter(x => x).length, 0)
}

const sideB = (arr) => {
  const followInstruction = (grid, instruction) => {
    for (let x = instruction.start.x; x <= instruction.end.x; ++x) {
      for (let y = instruction.start.y; y <= instruction.end.y; ++y) {
        if (grid[x] === undefined) grid[x] = []
        if (grid[x][y] === undefined) grid[x][y] = 0
        switch (instruction.action) {
          case 'turn on':
            grid[x][y] = grid[x][y] + 1
            break
          case 'turn off':
            grid[x][y] = grid[x][y] > 0 ? grid[x][y] - 1 : 0
            break
          case 'toggle':
            grid[x][y] = grid[x][y] + 2
            break
        }
      }
    }
    return grid
  }

  return arr.reduce(followInstruction, [])
    .reduce((sum, row) => sum + row.reduce((a, b) => a + (b || 0)), 0)
}

const main = () => {
  const contents = fs.readFileSync('../6.txt', 'utf8')
  const input = parse(contents)
  console.log(sideA(input))
  console.log(sideB(input))
}

main()

File: r/6.R
#!/usr/bin/env Rscript

parse <- function(contents) {
  parseLine <- function(line) {
    words <- unlist(strsplit(line, ' '))

    if (words[1] == 'turn') {
      instruction <- paste(words[1], words[2], sep = '')
      optind <- 3
    } else {
      instruction <- words[1]
      optind <- 2
    }

    x0 <- as.numeric(unlist(strsplit(words[optind], ','))[1]) + 1
    x1 <- as.numeric(unlist(strsplit(words[optind+2], ','))[1]) + 1

    y0 <- as.numeric(unlist(strsplit(words[optind], ','))[2]) + 1
    y1 <- as.numeric(unlist(strsplit(words[optind+2], ','))[2]) + 1

    data.frame(instruction, x0, x1, y0, y1)
  }

  do.call(rbind, lapply(contents, parseLine))
}

side.a <- function(input) {
  lights <- matrix(data = FALSE, nrow = 1000, ncol = 1000)

  follow.instruction <- function(r) {
    type <- r[1]
    x0 <- r[2]
    x1 <- r[3]
    y0 <- r[4]
    y1 <- r[5]

    if (type == 'turnon') {
      lights[y0:y1, x0:x1] <<- TRUE
    } else if (type == 'turnoff') {
      lights[y0:y1, x0:x1] <<- FALSE
    } else if (type == 'toggle') {
      lights[y0:y1, x0:x1] <<- !lights[y0:y1, x0:x1]
    }
  }

  apply(input, 1, follow.instruction)
  sum(lights)
}

side.b <- function(input) {
  lights <- matrix(data = 0, nrow = 1000, ncol = 1000)

  follow.instruction <- function(r) {
    type <- r[1]
    x0 <- r[2]
    x1 <- r[3]
    y0 <- r[4]
    y1 <- r[5]

    if (type == 'turnon') {
      lights[y0:y1, x0:x1] <<- lights[y0:y1, x0:x1] + 1
    } else if (type == 'turnoff') {
      lights[y0:y1, x0:x1] <<- lights[y0:y1, x0:x1] - 1
    } else if (type == 'toggle') {
      lights[y0:y1, x0:x1] <<- lights[y0:y1, x0:x1] + 2
    }
    lights[lights < 0] <<- 0
  }

  apply(input, 1, follow.instruction)
  sum(lights)
}

main <- function() {
    contents <- readLines('../6.txt')
    input <- parse(contents)
    cat(side.a(input), '\n')
    cat(side.b(input), '\n')
}

main()
